stacks-common/src/util/lru_cache.rs

```diff
@@ -0,0 +1,255 @@
+// Copyright (C) 2024 Stacks Open Internet Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use hashbrown::HashMap;
+
+/// Node in the doubly linked list
+struct Node<K, V> {
+    key: K,
+    value: V,
+    dirty: bool,
+    next: usize,
+    prev: usize,
+}
+
+/// LRU cache for account nonces
+pub struct LruCache<K, V> {
+    capacity: usize,
+    /// Map from address to an offset in the linked list
+    cache: HashMap<K, usize>,
+    /// Doubly linked list of values in order of most recently used
+    order: Vec<Node<K, V>>,
+    /// Index of the head of the linked list -- the most recently used element
+    head: usize,
+    /// Index of the tail of the linked list -- the least recently used element
+    tail: usize,
+}
+
+impl<K: Eq + std::hash::Hash + Clone, V: Copy> LruCache<K, V> {
+    /// Create a new LRU cache with the given capacity
+    pub fn new(capacity: usize) -> Self {
+        LruCache {
+            capacity,
+            cache: HashMap::new(),
+            order: Vec::with_capacity(capacity),
```

Comment:
@jbencin What do you think about this? Is it worth setting the capacity here or better to just let it grow? I suspect that the default capacity can be greatly reduced from what it is currently set to (`1024 * 1024`). This is where it would help to have some good benchmarking.

Reply:
Well first of all, you can't let cache grow unbounded, or that could lead to a DoS attack. So you have to set a limit. And you might as well allocate the entire limit up front, since you don't want to find out after running a node for a week that you're OOM because your cache got too big. Pre-allocation also lets you avoid latency spikes from large `realloc()`s

So yeah, I say use `with_capacity()` and never resize it

Reply:
Right, I meant still capping the capacity, but just not allocating it all at once. Sounds good to just leave it as-is.